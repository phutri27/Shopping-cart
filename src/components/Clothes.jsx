import { useParams, useRouteLoaderData } from "react-router";
import { useCart } from "./Cart";
import Header from "./Header";
function Clothes(){
    const {id} = useParams();
    const datas = useRouteLoaderData("rootman");
    const result = datas.find(data => data.id === Number(id))
    const { setQuantity } = useCart();

    function addToCart(){
        setQuantity(q => q + 1)
    }

    return (
        <div>
            <Header />
            <img src={result.imageUrl} alt={result.name} />
            <button onClick={addToCart}>ADD TO CART</button>
        </div>
    )
}

export default Clothes