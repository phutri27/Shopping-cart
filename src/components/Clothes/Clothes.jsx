import { useParams, useRouteLoaderData } from "react-router";
import { useCart } from "../CartQuantity";
import Header from "../Header";
import styles from "./clothes.module.css"

function Clothes({head}){
    const {id} = useParams();
    const datas = useRouteLoaderData(head);
    let result = datas.find(data => data.id === Number(id))
    const { setItems, items } = useCart();

    function addToCart(){
        const match = items.find(item => item.id === result.id)
        if (!match || match === undefined){
            result.quantity = 1;    
            setItems(it => [...it, result])
        }
        else{
            setItems(item => item.map(it => (
                it.id === result.id ? {...it, quantity: it.quantity + 1} : it
            )))
        }
    }

    return (
        <div>
            <Header />
            <div className={styles.layout}>
                <img src={result.imageUrl} alt={result.name} />
                <div className={styles.inner}>
                    <div className={styles.title}>
                        <p className={styles.name}>{result.name.toUpperCase()}</p>
                        <p className={styles.text}>{result.price.current.text}</p>
                    </div>
                    <div className={styles.price}>
                        <p>COLOR: {result.colour}</p>
                        <img src={result.imageUrl} alt={result.name} />
                    </div>
                    <div className={styles.buttonDiv}>
                        <button onClick={addToCart}>ADD TO CART</button>
                    </div>
                </div>
            </div>
        </div>
    )
}

export default Clothes  