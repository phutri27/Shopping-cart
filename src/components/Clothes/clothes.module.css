.layout {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  gap: 10vw;
  font-family: "Noto Sans", sans-serif;
}
.layout img {
  transform: scale(1.3);
}
.layout .inner {
  height: 70vh;
  display: flex;
  flex-direction: column;
}
.layout .inner .title .text {
  font-weight: bold;
}
.layout .inner .title .name {
  font-weight: 500;
}
.layout .inner .price {
  font-size: 0.8rem;
  display: flex;
  flex-direction: column;
  overflow: visible;
  font-weight: 600;
}
.layout .inner .price img {
  margin: 10px 0 0 10px;
  width: 61px;
  height: 90px;
  border: 1px solid black;
}
.layout .inner .buttonDiv {
  display: flex;
  flex: 3.5;
  align-items: center;
  justify-content: center;
}
.layout .inner .buttonDiv button {
  font-family: "Noto Sans", sans-serif;
  font-size: 1.05rem;
  color: white;
  background-color: black;
  border: none;
  width: 100%;
  padding: 20px 0px;
  cursor: pointer;
  transition: opacity 0.15s;
}
.layout .inner .buttonDiv button:hover {
  opacity: 0.7;
}

/*# sourceMappingURL=clothes.module.css.map */
