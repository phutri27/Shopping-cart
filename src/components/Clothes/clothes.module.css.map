{"version":3,"sourceRoot":"","sources":["clothes.module.scss"],"names":[],"mappings":"AAAA;EACI;EACA;EACA;EACA;EACA;EACA;;AACA;EACI;;AAEJ;EACI;EACA;EACA;;AAEI;EACI;;AAEJ;EACI;;AAGR;EACI;EACA;EACA;EACA;EACA;;AACA;EACI;EACA;EACA;EACA;;AAGR;EACI;EACA;EACA;EACA;;AACA;EACI;EACA;EACA;EACA;EACA;EACA;EACA;EACA;EACA;;AACA;EACI","file":"clothes.module.css"}