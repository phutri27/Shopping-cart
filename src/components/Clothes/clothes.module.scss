.layout{
    display: flex;
    justify-content: center;
    align-items: center;
    height: 100vh;
    gap: 10vw;
    font-family: "Noto Sans", sans-serif;   
    img{
        transform: scale(1.3);
    }
    .inner{
        height: 70vh;
        display: flex;
        flex-direction: column;
        .title{
            .text{
                font-weight: bold;
            }
            .name{
                font-weight: 500;
            }
        }
        .price{
            font-size: 0.8rem;
            display: flex;
            flex-direction: column;
            overflow:visible; 
            font-weight: 600;
            img{
                margin: 10px 0 0 10px;
                width: 61px;
                height: 90px;
                border: 1px solid black;
            }
        }
        .buttonDiv{
            display: flex;
            flex: 3.5;
            align-items: center;
            justify-content: center;
            button{
                font-family: "Noto Sans", sans-serif;
                font-size: 1.05rem;
                color: white;
                background-color: black;
                border: none;
                width: 100%;
                padding: 20px 0px;
                cursor: pointer;
                transition: opacity 0.15s;
                &:hover{
                    opacity: 0.7;
                }
            }
        }
        
    }
}