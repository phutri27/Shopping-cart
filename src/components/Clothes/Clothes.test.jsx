import React from "react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useParams, useRouteLoaderData } from "react-router";
import { useCart } from "../CartQuantity";
import Clothes from "./Clothes";

vi.mock("react-router", () => ({
  useParams: vi.fn(),
  useRouteLoaderData: vi.fn(),
}));

vi.mock("../Header", () => ({
  default: () => <div data-testid="header">Header</div>,
}));

vi.mock("../CartQuantity", () => ({
  useCart: vi.fn(),
}));

const sampleData = [
  {
    id: 1,
    name: "Blue Shirt",
    imageUrl: "https://example.com/shirt.jpg",
    price: { current: { text: "$20.00", value: 20 } },
    colour: "Blue",
  },
  {
    id: 2,
    name: "Red Dress",
    imageUrl: "https://example.com/dress.jpg",
    price: { current: { text: "$40.00", value: 40 } },
    colour: "Red",
  },
];

describe("Clothes component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("renders correct product data from route", () => {
    // arrange
    useParams.mockReturnValue({ id: "1" });
    useRouteLoaderData.mockReturnValue(sampleData);
    useCart.mockReturnValue({ items: [], setItems: vi.fn() });


    // assert
    expect(render(<Clothes head="man" />)).toMatchSnapshot();
  });

    it("adds a new item to the cart if not present", async () => {
    const user = userEvent.setup();
    const mockSetItems = vi.fn();

    useParams.mockReturnValue({ id: "1" });
    useRouteLoaderData.mockReturnValue(sampleData);
    useCart.mockReturnValue({ items: [], setItems: mockSetItems });

    render(<Clothes head="man" />);

    const button = screen.getByRole("button", { name: /add to cart/i });
    await user.click(button);

    // should call setItems with an array containing the product
    expect(mockSetItems).toHaveBeenCalledTimes(1);
    const callArg = mockSetItems.mock.calls[0][0];
    expect(typeof callArg).toBe("function"); // because setItems is called with updater fn
  });
})
