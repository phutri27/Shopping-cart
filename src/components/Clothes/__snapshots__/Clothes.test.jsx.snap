// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Clothes component > renders correct product data from route 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          data-testid="header"
        >
          Header
        </div>
        <div
          class="_layout_f85ef2"
        >
          <img
            alt="Blue Shirt"
            src="https://example.com/shirt.jpg"
          />
          <div
            class="_inner_f85ef2"
          >
            <div
              class="_title_f85ef2"
            >
              <p
                class="_name_f85ef2"
              >
                BLUE SHIRT
              </p>
              <p
                class="_text_f85ef2"
              >
                $20.00
              </p>
            </div>
            <div
              class="_price_f85ef2"
            >
              <p>
                COLOR: 
                Blue
              </p>
              <img
                alt="Blue Shirt"
                src="https://example.com/shirt.jpg"
              />
            </div>
            <div
              class="_buttonDiv_f85ef2"
            >
              <button>
                ADD TO CART
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        data-testid="header"
      >
        Header
      </div>
      <div
        class="_layout_f85ef2"
      >
        <img
          alt="Blue Shirt"
          src="https://example.com/shirt.jpg"
        />
        <div
          class="_inner_f85ef2"
        >
          <div
            class="_title_f85ef2"
          >
            <p
              class="_name_f85ef2"
            >
              BLUE SHIRT
            </p>
            <p
              class="_text_f85ef2"
            >
              $20.00
            </p>
          </div>
          <div
            class="_price_f85ef2"
          >
            <p>
              COLOR: 
              Blue
            </p>
            <img
              alt="Blue Shirt"
              src="https://example.com/shirt.jpg"
            />
          </div>
          <div
            class="_buttonDiv_f85ef2"
          >
            <button>
              ADD TO CART
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
